lib/creditCards.js

```diff
@@ -0,0 +1,324 @@
+    ],
+    length: [
+      [16]
+    ],
+    validate: true
+  },
+  {
+    key: 'laser',
+    name: 'Laser',
+    iin: [
+      [6304],
+      [6706],
+      [6771],
+      [6709]
+    ],
+    length: [
+      [16, 19]
+    ],
+    validate: true
+  },
+  {
+    key: 'maestro',
+    name: 'Maestro',
+    iin: [
+      [5018],
+      [5020],
+      [5038],
+      [5612],
+      [5893],
+      [6304],
+      [6759],
+      [6761],
+      [6762],
+      [6763],
+      ['0604'],
+      [6390]
+    ],
+    length: [
+      [12, 19]
+    ],
+    validate: true
+  },
+  {
+    key: 'dankort',
+    name: 'Dankort',
+    iin: [
+      [5019]
+    ],
+    length: [
+      [11],
+      [16]
+    ],
+    validate: true
+  },
+  {
+    key: 'mastercard',
+    name: 'MasterCard',
+    iin: [
+      [50, 55]
+    ],
+    length: [
+      [16]
+    ],
+    validate: true
+  },
+  {
+    key: 'solo',
+    name: 'Solo',
+    iin: [
+      [6334],
+      [6767]
+    ],
+    length: [
+      [16],
+      [18, 19]
+    ],
+    validate: true
+  },
+  {
+    key: 'switch',
+    name: 'Switch',
+    iin: [
+      [4903],
+      [4905],
+      [4911],
+      [4936],
+      [564182],
+      [633110],
+      [6333],
+      [6759]
+    ],
+    length: [
+      [16],
+      [18, 19]
+    ],
+    validate: true
+  },
+  {
+    key: 'visa',
+    name: 'Visa',
+    iin: [
+      [4]
+    ],
+    length: [
+      [13],
+      [16]
+    ],
+    validate: true
+  },
+  {
+    key: 'visa_electron',
+    name: 'Visa Electron',
+    iin: [
+      [4026],
+      [417500],
+      [4405],
+      [4508],
+      [4844],
+      [4913],
+      [4917]
+    ],
+    length: [
+      [16]
+    ],
+    validate: true
+  }
+]
+
+// Card matching
+function getCardType (type) {
+  return find(cardTypes, { key: type })
+}
+
+function getLikelyMatch (num) {
+  const matches = compact(getPossibleMatches(num))
+  let key = ''
+  let val = 0
+  forEach(matches, function (d) {
+    if (d.priority > val) {
+      key = d.key
+      val = d.priority
+    }
+  })
+  return key
+}
+
+function getPossibleMatches (num) {
+  return map(cardTypes, function (d) {
+    var match = flatten(isMatch(d.iin, num))
+    if (match.length) { return { key: d.key, priority: match[0] } }
+  })
+}
+
+function isMatch (arr, num) {
+  return filter(arr, function (d) {
+    if (d.length > 1) { return isInRange(d, num) }
+    return num.indexOf(d) === 0
+  })
+}
+
+function isInRange (arr, num) {
+  return !!filter(range(arr[0], arr[1] + 1), function (d) {
+    return num.indexOf(d) === 0
+  }).length
+}
+
+// Card validation methods
+function validate (num) {
+  var type = getLikelyMatch(num)
+  if (!type || !isValidLength(type, num)) { return false }
+  return !getCardType(type).validate || luhnTest(num)
+}
+
+function isValidLength (type, num) {
+  return !!isMatch(getCardType(type).length, num.length.toString()).length
+}
+
+function luhnTest (a, b, c, d, e) {
+  for (d = +a[b = a.length - 1], e = 0; b--;) {
```

Comment:
Yes, and lets go back to writing:

```
for (var i = items.length; i > 0; i--) {}
```

Because it's more performant than `forEach`...

Or, we could optimise for legibility except when we have evidence that there is visible performance impact...